content/blog/debugging-dde.md

```diff
@@ -0,0 +1,149 @@
+except one. The test that produced different output for 32-bit than for 
+64-bit was about the simplest case possible: a function that always returns `1`.
+
+```
+  deriv <- function(t, y, p) { 1 }
+  dde::dopri(0, c(0, 1), deriv, 0)
+```
+
+For the two timesteps, `0` and `1`, on 64-bit, it correctly returned `(1,1)`; 
+on 32-bit, it returned `(1,0)`.
+
+## Primary Approaches
+
+At first, we wondered if it was a memory issue that only surfaced on 32-bit. 
+So we checked beyond doubt that `sizeof(size_t)` is `4` on 32-bit, but `8` on 
+64-bit, as are all pointers. All good. Other types were the same for all 
+platforms. But there were no obvious violations; Valgrind on linux picked up
+nothing.
+
+Resorting to random black-box bashing: `dde:dopri(0, c(0, 1.121), deriv, 0)`
+and smaller numbers failed the test. `dde:dopri(0, c(0, 1.122), deriov, 0)` and
+larger numbers all passed. Eureka? Nope - a mysterious red herring, alas.
+
+In the real-life code, the `(a>0)` and `(b>0)` were a bit more complicated -
+but not much; no functions involved, just comparisons between members of a
+struct, so no sensible way they could exit the loop, and then appear
+to have met the criteria for looping in retrospect, as we were seeing. 
+
+## Secondary Flailings
+
+The next level of desperation involved dumping every member of the object at
+intervals with `Rprintf`. This revealed some corruption at the end of the code, 
+surely evidence of some bad memory writes! Wrong again - they were tracked down
+to different, but very acceptable behaviour as the object had already been 
+freed and deconstructed by then, so could rightly return undefined values.
+
+But this did expose a genuine difference between platforms in a variable; 
+narrowed down to the `while` loop above, which terminated early (or didn't 
+get entered at all), under certain 32-bit circumstances. Printing things within
+the loop - or sometimes before it - sometimes changed the behaviour and made
+things work, but it would be poor form to submit a PR that claims to fix a functional
+problem by printing a ton of waffle before-hand.
+
+My github commit titles around this time are telling. **Total Confusion Reigns** 
+towards **Wits' End** and later a  **Deeper Confusion** that's somehow worse than the 
+previous *total* one. Eventually, as hope ebbed away, the crucial summit arrived 
+at a commit aptly named **Another bizarre effort**.
+
+The question now moves from "why *doesn't* it work" to, "why *does* it":-
+
+```
+cond = (a>0) && (b>0);
+while (((a>0) && (b>0)) || cond) {
+  // a and b may get changed
+  cond = (a>0) && (b>0);
+}
+
+```
+The syntax is surely equivalent to the code at the top, if 
+more wordy with some redundency. A good PR might recommend
+cleaning it up back into the code at the top. Nevertheless,
+looking at `a` and `b` twice for each decision caused correct entry and exit
+of the loop, and the appveyor tests passed fully.
```

Comment:
it will not be clear to most that `appveyor` is a windows CI service

Reply:
Edited earlier when I first mention it...